lib/handler/mruby/middleware.c

```diff
@@ -0,0 +1,722 @@
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_socktype = SOCK_STREAM;
+    hints.ai_protocol = IPPROTO_TCP;
+    hints.ai_flags = AI_ADDRCONFIG | AI_NUMERICSERV;
+    if (getaddrinfo(hostname, servname, &hints, &res) != 0)
+        goto Error;
+
+    switch(res->ai_family) {
+        case AF_INET:
+        case AF_INET6:
+            memcpy(ss, res->ai_addr, res->ai_addrlen);
+            break;
+        default:
+            goto Error;
+    }
+
+    socklen_t len =  res->ai_addrlen;
+    freeaddrinfo(res);
+    return len;
+
+Error:
+    if (res != NULL)
+        freeaddrinfo(res);
+    mrb->exc = NULL;
+    return 0;
+}
+
+#define KEY_PREFIX "HTTP_"
+#define KEY_PREFIX_LEN (sizeof(KEY_PREFIX) - 1)
+
+static h2o_iovec_t convert_env_to_header_name(h2o_mem_pool_t *pool, const char *name, size_t len)
+{
+    if (len < KEY_PREFIX_LEN || ! h2o_memis(name, KEY_PREFIX_LEN, KEY_PREFIX, KEY_PREFIX_LEN)) {
+        return h2o_iovec_init(NULL, 0);
+    }
+
+    h2o_iovec_t ret;
+
+    ret.len = len - KEY_PREFIX_LEN;
+    ret.base = h2o_mem_alloc_pool(pool, ret.len);
+
+    name += KEY_PREFIX_LEN;
+    char *d = ret.base;
+    for (; len != 0; ++name, --len)
+        *d++ = *name == '_' ? '-' : h2o_tolower(*name);
+
+    return ret;
+}
+
+#undef KEY_PREFIX
+#undef KEY_PREFIX_LEN
+
+static int handle_request_header(h2o_mruby_shared_context_t *shared_ctx, h2o_iovec_t name, h2o_iovec_t value, void *_req)
+{
+    h2o_req_t *req = _req;
+    const h2o_token_t *token;
+
+    /* convert env key to header name (lower case) */
+    name = convert_env_to_header_name(&req->pool, name.base, name.len);
+    if (name.base == NULL)
+        return 0;
+
+    if ((token = h2o_lookup_token(name.base, name.len)) != NULL) {
+        if (token == H2O_TOKEN_TRANSFER_ENCODING) {
+            /* skip */
+        } else if (token == H2O_TOKEN_CONTENT_LENGTH) {
+            req->content_length = h2o_strtosize(value.base, value.len);
+        } else {
+            value = h2o_strdup(&req->pool, value.base, value.len);
+            h2o_add_header(&req->pool, &req->headers, token, NULL, value.base, value.len);
+        }
+    } else {
+        value = h2o_strdup(&req->pool, value.base, value.len);
+        h2o_add_header_by_str(&req->pool, &req->headers, name.base, name.len, 0, NULL, value.base, value.len);
+    }
+
+    return 0;
+}
+
+static void on_subreq_error_callback(h2o_req_t *req, void *_data, h2o_iovec_t error)
+{
+    struct st_mruby_subreq_t *subreq = (void *)_data;
+    mrb_state *mrb = subreq->ctx->shared->mrb;
+
+    assert(!mrb_nil_p(subreq->error_stream));
+
+    mrb_value msg = mrb_str_new(mrb, error.base, error.len);
+    mrb_funcall(mrb, subreq->error_stream, "write", 1, msg);
+}
+
+static struct st_mruby_subreq_t *create_subreq(h2o_mruby_context_t *ctx, mrb_value env)
+{
+    mrb_state *mrb = ctx->shared->mrb;
+    int i;
+    int gc_arena = mrb_gc_arena_save(mrb);
+    mrb_gc_protect(mrb, env);
+
+    mrb_value dupenv = mrb_funcall(mrb, env, "dup", 0);
```

Comment:
I would appreciate it if you could do `env = mrb_funcall(..., env, ...)` to replace `env`.
Having both `dupenv` and `env` available as different variables in this big function seems scary to me.